File: 4/palindrome.cpp
//A palindromic number reads the same both ways.
// The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
//Find the largest palindrome made from the product of two 3-digit numbers.

#include <iostream>
#include <vector>
#include <string>

using namespace std;

bool isPalindrome(const string& word, int begin, int end) {
    if(begin >= end) return true;
    if(word[begin]==word[end]) return isPalindrome(word, begin+1, end-1);
    return false;
}

//finds palindromes made from the product of two 3-digit numbers (100-999)
void findPalindrome() {
    string temp;
    int largest = 0;
    for(int i = 100; i <= 999; ++i) {
        for(int j = 100; j <= 999; ++j) {
            temp = to_string(i*j);
            if(isPalindrome(temp, 0, static_cast<int>(temp.length()-1))) {
                if(i*j > largest) largest = i*j;
//                printf("%d * %d = %d :: IS A PALINDROME\n",i,j,i*j);
            }
        }
    }
    printf("%d is the largest PALINDROME", largest);
}

int main() {
//    string foo = to_string(5005005);
//    if(isPalindrome(foo, 0, static_cast<int>(foo.length()-1))){
//        cout << "YES";
//    }
    findPalindrome();

    return 0;
}

File: 4/palindrome.py
print("""
Largest Palindrome product
Problem 4

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91x 99.

Find the largest palindrome made from the product of two 3-digit numbers.

Answer: 906609

""")

from time import clock

def _isPalindrome(word, start, end):
    if start >= end: return True
    if word[start] == word[end]: return _isPalindrome(word, start+1, end-1)
    return False

def isPalindrome(word):
    return _isPalindrome(word, 0, len(word)-1)

while True:
    try:
        limit = int(input("Enter limit (Default 999): "))
    except ValueError:
        limit = 1000
    start = clock()
    myList = [int(i*j) for i in range(int(limit/10),limit) for j in range(i,limit) if isPalindrome(str(i*j))]
    print("The largest palindromic number of numbers less than {} is {}".format(limit, max(myList)))
    print("The program took {} seconds to run".format(clock()-start))
